File: guests/haskell/sections.hs
-- The sections scene, Haskell port — guests/rust/sections.rs,
-- tools/scenes/sections.steps.

import Data.IORef (modifyIORef', newIORef, readIORef)
import Data.Word (Word64)
import KayaApp
import KayaWire (Value (..))

feedId, archiveId :: Word64
feedId = 7
archiveId = 8

-- The SIDEBAR half rides an AUX WINDOW opened only from the desktop tail's
-- click, so 'createWindow' never runs where the capability is absent.
libraryId, shelvesId, loansId :: Word64
libraryId = 1
shelvesId = 2
loansId = 3

main :: IO ()
main = kayaMain $ \app -> do
  visitTally <- newIORef (0 :: Int)
  _ <- buildTx app $ do
    window 0 [WTitle "sections", WSectionsPresentation 1]
    visits <- signal (VStr "archive: 0 visits")
    -- A symbol names a CONCEPT (docs/styling-plan.md D6).
    addSection feedId [STitle "Feed", SSymbol SymbolHome]
    addSection
      archiveId
      [ STitle "Archive",
        SSymbol SymbolStar,
        SOnSelected
          ( do
              modifyIORef' visitTally (+ 1)
              n <- readIORef visitTally
              buildTx app $
                writeSignal visits (VStr ("archive: " ++ show n ++ " visits"))
          )
      ]
    feedRoot <-
      column
        []
        [ do
            ready <- signal (VStr "feed ready")
            labelBound ready, -- label#0
          buttonOn "to archive" $
            -- Programmatic selection does NOT echo: 'SOnSelected' must not fire.
            buildTx app (selectSection archiveId), -- button#0
          buttonOn "open library" $ -- button#1
            buildTx app $ do
              createWindow libraryId [WTitle "library", WSectionsPresentation 2]
              addSectionIn libraryId shelvesId [STitle "Shelves", SSymbol SymbolSearch]
              addSectionIn libraryId loansId [STitle "Loans", SSymbol SymbolLock]
              shelvesRoot <-
                column
                  []
                  [ do
                      ready <- signal (VStr "shelves ready")
                      labelBound ready -- label#2
                  ]
              mountIn shelvesId shelvesRoot
              loansRoot <-
                column
                  []
                  [ do
                      ready <- signal (VStr "loans ready")
                      labelBound ready -- label#3
                  ]
              mountIn loansId loansRoot
        ]
    mountIn feedId feedRoot
    archiveRoot <- column [] [labelBound visits] -- label#1
    mountIn archiveId archiveRoot
  return ()

File: guests/rust/sections.rs
//! The sections conformance scene (tools/scenes/sections.steps): a user's
//! switch emits, a programmatic select_section does not.

use kaya::WindowId;

#[derive(Clone)]
enum Msg {
    ArchiveShown,
    GoArchive,
    OpenLibrary,
}

const FEED: WindowId = WindowId(7);
const ARCHIVE: WindowId = WindowId(8);
// An AUX WINDOW, reached only by the desktop tail's click, so create_window
// never runs where the capability is absent.
const LIBRARY: WindowId = WindowId(1);
const SHELVES: WindowId = WindowId(2);
const LOANS: WindowId = WindowId(3);

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let visits_label = ctx.apply(|tx| {
        // The presentation hint is ADVISORY: no observable rides on it.
        tx.window(kaya::DEFAULT_WINDOW)
            .title("sections")
            .sections_presentation(kaya::SectionsPresentation::Bar);
        // SF Symbols are licensed to Apple platforms: no shared asset.
        let feed = tx.add_section(FEED).title("Feed").symbol(kaya::Symbol::Home).id();
        let archive = tx
            .add_section(ARCHIVE)
            .title("Archive")
            .symbol(kaya::Symbol::Star)
            .id();
        msgs.on_section_selected(archive, Msg::ArchiveShown);

        let feed_root = tx
            .column(|tx| {
                let ready = tx.signal("feed ready");
                tx.label(ready); // label#0
                let go = tx.button("to archive").id(); // button#0
                msgs.on_click(go, Msg::GoArchive);
                let open = tx.button("open library").id(); // button#1
                msgs.on_click(open, Msg::OpenLibrary);
            })
            .id();
        tx.mount_in(feed, feed_root);

        let visits = tx.signal("archive: 0 visits");
        let archive_root = tx
            .column(|tx| {
                tx.label(visits); // label#1
            })
            .id();
        tx.mount_in(archive, archive_root);
        visits
    });

    let mut visits = 0u64;
    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::ArchiveShown => {
                visits += 1;
                ctx.apply(|tx| {
                    tx.write(visits_label, format!("archive: {visits} visits"));
                });
            }
            Msg::GoArchive => {
                ctx.apply(|tx| {
                    tx.select_section(ARCHIVE);
                });
            }
            Msg::OpenLibrary => {
                ctx.apply(|tx| {
                    tx.create_window(LIBRARY)
                        .title("library")
                        .sections_presentation(kaya::SectionsPresentation::Sidebar);
                    let shelves = tx
                        .add_section_in(LIBRARY, SHELVES)
                        .title("Shelves")
                        .symbol(kaya::Symbol::Search)
                        .id();
                    let loans = tx
                        .add_section_in(LIBRARY, LOANS)
                        .title("Loans")
                        .symbol(kaya::Symbol::Lock)
                        .id();
                    let shelves_root = tx
                        .column(|tx| {
                            let l = tx.signal("shelves ready");
                            tx.label(l); // label#2
                        })
                        .id();
                    tx.mount_in(shelves, shelves_root);
                    let loans_root = tx
                        .column(|tx| {
                            let l = tx.signal("loans ready");
                            tx.label(l); // label#3
                        })
                        .id();
                    tx.mount_in(loans, loans_root);
                });
            }
        }
    }
}

fn main() {
    kaya::run(app)
}
